### ISSUE TITLE
[BUG]: Updating Docker images with a "base", that updates docker environment variables, doesn't change upon updating

### ISSUE THREAD

**Issue by engels74** [azukaar/Cosmos-Server#246]
### What happened?

So there's this issue I had, specifically with the Docker images that [hotio](https://hotio.dev) supplies.

He recently updated his [base image](https://github.com/hotio/base/tree/alpinevpn), with a change to a Docker ENV/VAR, named `S6_STAGE2_HOOK`. It was [changed](https://github.com/hotio/base/commit/c0b11d5031246144bdefb04ac8cbf7479608c7ae) from `/init-hook` to `/etc/s6-overlay/init-hook`.

When the images were updated in Cosmos-server, they didn't change the existing images environment variables to reflect this. They were left as `/init-hook`, which broke most of his images.

### What should have happened?

The variable `S6_STAGE2_HOOK` should've been updated on the existing containers.

### How to reproduce the bug?

1. Download any hotio release older than [this commit](https://github.com/hotio/base/commit/c0b11d5031246144bdefb04ac8cbf7479608c7ae)
2. E.g. this radarr release [here](https://github.com/orgs/hotio/packages/container/radarr/206358088?tag=release-efa66c3)
3. Run the container
4. Check if the `S6_STAGE2_HOOK` has the value of `/init-hook`
5. Update the container / Pull New Image
6. Check Cosmos, and see that `S6_STAGE2_HOOK` is still set to `/init-hook`
7. Check logs, and see that everything has gone haywire
8. Remove the var `S6_STAGE2_HOOK` manually through Cosmos
9. Recreate the container
10. The now-correct (updated) `S6_STAGE2_HOOK` will show up with `/etc/s6-overlay/init-hook`

### Other details

The easy solution was to manually go into every Docker container through Cosmos, remove the `S6_STAGE2_HOOK` variable, press "Update Container" and let Cosmos recreate it, with the now-correct `S6_STAGE2_HOOK` variable.

I'm thinking this would be great if Cosmos would take this into account, when updating images. That being, when "base" images are updated, and if that somehow affects or changes existing containers variables. So that the user can use the auto update, and be sure knowing that Cosmos will update existing variables ;).

A simple GIF of how it was "solved" in Cosmos:
https://i.imgur.com/hTJKwks.gif

I'm sure other users will end up having this problem, atleast if they use hotio's containers. I made a neat little simple bash script to detect what containers has the `S6_STAGE2_HOOK` variable set to `/init-hook`:
<https://logs.notifiarr.com/?37e0ff883c53d517=#8WuwvRSycAV4MzcMJJsDLDYtEtT48CmUbryiU6C8geUz>

### System details

- OS: Ubuntu 22.04.4 LTS (running in VM on UnraidOS 6.12.10)
- Browser: Brave
- Version: Cosmos Server v0.15.5

**Comment by azukaar** [comment]
Thanks for the throughout ticket! Really appreciate it
The main issue with this is that it's difficult to know what var the user has edited or not, the "overwriting" vars is normally a Docker thing (where you destroy the container, and re-create it without the var that the container has set in the first place), but because from the UI you update containers including the var the container has created, so Docker loses track of which one came from the container

**Comment by engels74** [comment]
No problem 😁 

Portainer apparently has the same issue, according to the hotio discord. Is there anything to do about it? 🤔

**Comment by azukaar** [comment]
Yeah I'd imagined so. Usually I try to fix even issues Portainer have in Cosmos, like this one : https://github.com/portainer/portainer/issues/2657

But for this, I am really not sure what to do tbh... I think dev should not do this kind of stuff in the first place honestly, env var are here to document a deployement's specific settings. if something is an env var it means we should be fully prepared for users to customize it in the first place, that does not look like it's even possible here because the value is then hardcoded in the RUN command

The only thing that could be done is, if the app was in the market, updating the env var inside the compose in the market woudl actually fix the container if you updated it from the market, basically

**Comment by madejackson** [comment]
I just stumbled upon this issue and I'd like to add something to this.

This often breaks automatic updates, f.e.  java based application as far as I am aware need JAVA_HOME set pointing to the right jdk version. If this version gets bumped, the update kills the container until manual intervention. Using ENV's for a PATH decleration is quiet common.

The same issue applies to changing entrypoints and/or commands.

I think cosmos compose should behave more like docker compose, using declarative configs which wont get overwritten at runtime like it happens atm.

**Comment by azukaar** [comment]
@madejackson 
Essentially the issue is that Docker as a running container and docker-compose as files, are two separated things. The docker compose has a set of option, but the resulting container will be completely different.

The "point" of the Cosmos UI is to see the container as it is running, with its final env, labels, networks, etc... Rather than a theorical shape that might not reflect what's actually running making debugging harder.

Basically there's 2 ways you can do this
 - compose first like Dockge, but then you have no idea what your containers actually look like in practice in your system, and you have to use the CLI to debug, and edit the compose to somehow get what you want after trial and error
 - container first like Cosmos, where thigns are more transparent, but you lose the ability to know when a container edits itself vs. when the user does.

I have no good solution to that problem as of know, but flipping the design 180 degrees to run docker-compose I dont think is the solution

**Comment by madejackson** [comment]
I don't think it's necessary to flip it 180°. Coming from kubernetes (k3s+rancher), both at once is entirely possible. Basically you have the config you set at installation, stored on disk  / in a config file + the runtime config seperately.

In cosmos this could be something like this: 
- Every parameter defined at installation will be displayed and used for updating the container.
- Any parameter extracted from runtime will be displayed as well, but greyed out and not used for updating/recreating the container. As soon as someone edits a greyed out parameter, it gets colorized/enabled for next update. To grey it out again, just delete it completely.

**Comment by azukaar** [comment]
That is still a big change in the UX and mode of operation. It's also very confusing to have an "install" config and a "runtime" config from the perspective of the user (not everyone understand the immutable nature of containers)

The best middle ground I can think of, is to track user-edited labels and env in a label ex. `cosmos-user-env` and `cosmos-user-label` to nuke every other labels/envs on edit/recreate
